packages/@lwc/wds-core/src/node/plugins/mock-esm/resolve/mock-import.js

```diff
@@ -32,19 +36,59 @@ export const makeMockImportResolver =
     // mock: import, in the same file. I don't think web test
     // runner will wait for the previous import to resolve
     // before taking a look at the next one
-
+    const release = await mutex.acquire();
     const { resolvedImport, importExists } = await recursiveResolve({
       source: verbatimImport,
       context,
     });
 
     const mockControllerPath = `${MOCK_CONTROLLER_PREFIX}${resolvedImport}`;
     const moduleKey = withoutQs(resolvedImport);
-    mockedModules.set(moduleKey, {
-      mockControllerPath,
-      exportedNames,
-      importExists: !forceMock && importExists,
-    });
-
+    if (importExists) {
+      try {
+        const absolutePath = path.join(rootdir, resolvedImport);
+        const content = await fs.readFile(absolutePath, 'utf-8');
+        const [, _exports] = parseEsm(content);
+        const originalExportedNames = _exports.map((exp) => exp.n);
+        const invalidExports = exportedNames.filter(
+          (exportName) => !originalExportedNames.includes(exportName),
+        );
+        if (invalidExports.length > 0) {
+          throw new Error(`Invalid exports: ${invalidExports.join(', ')}`);
+        }
+        if (!mockedModules.has(moduleKey)) {
+          mockedModules.set(moduleKey, {
+            mockControllerPath,
+            exportedNames: originalExportedNames,
+            importExists: !forceMock && importExists,
+          });
+          // Further processing of the content can be done here
+        }
+      } catch (error) {
+        console.error('Error while mocking for resolved module:', error);
+      }
+    } else {
+      try {
+        if (mockedModules.has(moduleKey)) {
+          const existingEntry = mockedModules.get(moduleKey);
+          const newExportedNames = Array.from(
+            new Set([...existingEntry.exportedNames, ...exportedNames]),
+          );
+          mockedModules.set(moduleKey, {
+            ...existingEntry,
+            exportedNames: newExportedNames,
+          });
+        } else {
+          mockedModules.set(moduleKey, {
+            mockControllerPath,
+            exportedNames,
+            importExists: !forceMock && importExists,
+          });
+        }
+      } catch (error) {
+        console.error('Error while mocking for unresolved module:', error);
+      }
+    }
+    release();
```

Comment:
In a subsequent PR, if more work is done here in `mock-import.js`, I think it'd be worth splitting up this handler into two. It is starting to get somewhat large with lots of branching logic. It is fine for now.